### DISCUSSION TITLE
Requiring Composer to run plugin

### DISCUSSION THREAD

**Discussion by rubenvdlinde** [OpenWebconcept/Afspraken#16] [General]
Several Open Webconcept plugins (PDC, FORMS and PUB) require additional libraries that are loaded trough composer. This blocks the plugins from being installed on a “normal” WordPress installation  and block the usage of any WordPress container that is currently available form the WordPress community, bitname or docker community. In stead developers are forced to build a composer based WordPress themselves (which requires time and brakes the semantic versioning system) there is a loooooooong list of pro’s and con’s there. 

I would like to make a design decision to either
•	Follow the community guidelines and maintainers of the framework we use (WordPress)
•	Switch to a framework that uses composer (like symphony or zend)
•	Include libraries in the the plugins and omitting the whole composer dependency all together (minor code duplication)

**Comment by Htbaa** [comment]
Instead of requiring potential users to run the build process of a plugin themselves (Git clone/download zip from GitHub, run build tasks) every plugin should have an installable ZIP-release available. That's what I do with DataInMap. With every release I generate a new distfile which is a ZIP-file that's ready to be uploaded to a WordPress site.

Want to develop on it? Simply git clone and run the building steps. Nothing new there I'd say. I tend to provide a docker-compose file that provides a ready to use WordPress environment to work on that plugin. But I also include it in existing projects and simply use its path as a docker volume.

One thing that can (or will) be problematic is when multiple plugins use the same additional libraries because of duplicate functions and classes. I'm not sure about the community guidelines on including third-party libraries in a plugin, but I do now WordPress doesn't allow plugins in their directory that solely provide access to an additional library.

But after typing this maybe I don't quite understand the issue you're experiencing and trying to fix. Could you elaborate?

**Reply by Htbaa** [reply]
I just had a chat with @mvdhoek1 about PDC and now understand the issue you're describing (-:

I have zero experience in using a composer based WordPress, but after talking with Mike I understand you basically have a single `vendor/` directory in your website's root where ALL composer based plug-ins have their dependencies installed. Instead of each plug-in having its own `vendor/` folder.

When trying to install PDC base I ran into this issue with a normal WordPress installation. Even though I did a `composer install` and the dependencies were there, they weren't being loaded. Which in the end made sense because the `vendor/autoload.php` file wasn't called from the plug-in. A potential solution Mike and I discussed was to add a check to every plug-in if it has a `vendor/` folder and if so include the `vendor/autoload.php` file from the plug-in.

That way a plugin can be used both with traditional and composer based WordPress installations. It would still be preferable to supply distfiles which include the dependencies. This would make life a bit easier for the administrator of a traditional WordPress installation.

**Reply by sanderdekroon** [reply]
I wholeheartedly agree with this! I think it's a good idea to check if there is a local `vendor` folder and if there is, load that `autoload.php` file. 

I also agree with generating a self contained plugin zip _with_ dependencies on every tagged release.

**Comment by mvdhoek1** [comment]
I've just opened a [PR](https://bitbucket.org/openwebconcept/plugin-pdc-base/pull-requests/37/feat-require-composer-autoloadphp-when) in Bitbucket, the repository will be migrated in the nearby future but for now still in Bitbucket.

**Comment by mvdhoek1** [comment]
The addition has been added to v3.4.8.
